File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int solve(int max, TreeNode* root) {
        if(root == nullptr) {
            return 0;
        }
        if(root->val >= max) {
            return 1 + solve(root->val, root->left) + solve(root->val, root->right);
        }
        return solve(max, root->left) + solve(max, root->right);
    }
    int goodNodes(TreeNode* root) {
        return solve(INT_MIN, root);
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-0.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        left = [0] * n

        for i in range(n):
            if i == 0:
                left[i] = height[i]
            else:
                left[i] = max(left[i-1], height[i])
        
        right = 0
        ans = 0

        for i in range(n-1, -1, -1):
            if i == n-1:
                right = height[i]
            else:
                right = max(right, height[i])
            
            ans += min(left[i], right) - height[i]

        return ans